advection_test/functions.py

```diff
@@ -0,0 +1,1302 @@
+    - uc, vc: x- and y- winds on C-grid
+    - delp: pressure thickness of layer
+    - tracers: Dict{'smoke}
+    - density: (kg/m3 for air)
+    - dimensions: Dict{'nx', 'ny', 'nz', 'nx1', 'ny1', 'nxhalo', 'nyhalo'}
+    - units: Dict{'areaflux', 'courant', 'mass', 'wind'}
+    - origins: Dict{'compute_3d'}
+    - backend: 'numpy'
+    - fvt_dict: Dict{'grid_type', 'hord'}
+    - dt_acoustic: acoustic time step (default = 300 seconds)
+
+    Outputs:
+    - fluxPrep: Dict{'crx', 'cry', 'mfxd', 'mfyd', 'ucv', 'vcv', 'xaf', 'yaf'}
+        - crx, cry are Courant numbers
+        - ucv, vcv are the contravariant winds
+        - xaf, yaf are fluxes of area (dx*dy but incorporating wind speed)
+        - mfxd, mfyd are mass (volume) fluxes incorporating wind speed
+    - tracAdv: class instance that performs tracer advection
+    """
+
+    fluxPrep = run_finiteVolumeFluxPrep(
+        configuration,
+        uc,
+        vc,
+        delp,
+        density,
+        dimensions,
+        units,
+        origins,
+        backend,
+        dt_acoustic,
+    )
+
+    tracAdv = build_tracerAdvection(configuration, fvt_dict, tracers)
+
+    tracAdv_data = {
+        "tracers": tracers,
+        "delp": delp,
+        "mfxd": fluxPrep["mfxd"],
+        "mfyd": fluxPrep["mfyd"],
+        "crx": fluxPrep["crx"],
+        "cry": fluxPrep["cry"],
+    }
+
+    return tracAdv_data, tracAdv
+
+
+def run_advectionStepWithReset(
+    tracAdv_dataInit, tracAdv_data, tracAdv, dt, mpi_rank=None
+):
+    """
+    Use: tracAdv_data = run_advectionStepWithReset()
+
+    Inputs:
+    - tracAdv_dataInit: Dict{'tracers', 'delp', 'mfxd', 'mfyd', 'crx', 'cry'}
+                        of values to be reset to after each advection step
+    - tracAdv_data: Dict{'tracers', 'delp', 'mfxd', 'mfyd', 'crx', 'cry'}
+    - tracAdv: class instance of TracerAdvection
+    - dt: time step in seconds
+    - mpi_rank: if 0, prints differences from timestep, initial condition
+
+    Outputs:
+    - tracAdv_data: updated fields (of only tracer)
+    """
+
+    tmp = cp.deepcopy(tracAdv_data["tracers"])  # pre-advection tracer state
+
+    tracAdv(
+        tracAdv_data["tracers"],
+        tracAdv_data["delp"],
+        tracAdv_data["mfxd"],
+        tracAdv_data["mfyd"],
+        tracAdv_data["crx"],
+        tracAdv_data["cry"],
+        dt,
+    )
+
+    tracAdv_data["delp"] = cp.deepcopy(tracAdv_dataInit["delp"])
+    tracAdv_data["mfxd"] = cp.deepcopy(tracAdv_dataInit["mfxd"])
+    tracAdv_data["mfyd"] = cp.deepcopy(tracAdv_dataInit["mfyd"])
+    tracAdv_data["crx"] = cp.deepcopy(tracAdv_dataInit["crx"])
+    tracAdv_data["cry"] = cp.deepcopy(tracAdv_dataInit["cry"])
+
+    if mpi_rank == 0:
+        diff_timestep = tracAdv_data["tracers"]["smoke"].data - tmp["smoke"].data
+        diff_fromInit = (
+            tracAdv_data["tracers"]["smoke"].data
+            - tracAdv_dataInit["tracers"]["smoke"].data
+        )
+        print(
```

Comment:
should this be optional output?

Reply:
I guess that's what I tried to achieve by passing it rank - so it would only print out on rank 0. Would love a better way of doing this.